### DISCUSSION TITLE
uvcorn duplicates ASGI `root_path` when started with `--root-path`

### DISCUSSION THREAD

**Discussion by amol-** [Kludex/uvicorn#2490] [Potential Issue]
If ``--root-path`` is used, uvcorn will prefix every request path with it:

https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/http/h11_impl.py#L199-L200

This breaks ASGI web frameworks routing capabilities as a request for `/proxy/hello` would become `/proxy/proxy/hello` and thus, even if web frameworks strip the `root_path` (for example like Starlette does: https://github.com/encode/starlette/blob/342410d250687ec6bba16c20e3d309ae34826802/starlette/_utils.py#L97-L98 ) they would be unable to process the routing correctly.

When `--root-path=/proxy` is used it should probably be set into the request `"scope"`, but paths should not be prefixed with it to avoid duplicating the prefix.

**Comment by Kludex** [comment]
> When `--root-path=/proxy` is used it should probably be set into the request `"scope"` [...]

Where?

**Reply by amol-** [reply]
Sorry if I wasn't clear. The `root_path` is already correctly set in the `"scope"` (at https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/http/h11_impl.py#L212 )

But it's also prefixed to the request path itself ( https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/http/h11_impl.py#L199-L200 )

This causes an unexpected behaviour because the documentation states:

```
  --root-path TEXT                Set the ASGI 'root_path' for applications submounted below a given URL path.
```

The behaviour instead seems to be that `uvicorn` itself mounts the application in a subpath.

I guess a reasonable solution might be to have two separate options for those:
* `--root-path` which just only sets the `root_path`
* `--mountpoint` which actually serves the application in a subpath (both adjusting `path` and `root_path`)

**Reply by amol-** [reply]
PS: I'm happy to contribute a PR if we reach consensus over the behavior

**Comment by amol-** [comment]
An implementation has been proposed in https://github.com/encode/uvicorn/pull/2493
